pkg/jx/cmd/upgrade_apps_test.go

```diff
@@ -77,13 +80,372 @@ func TestUpgradeAppForGitOps(t *testing.T) {
 	assert.Equal(t, newVersion.String(), found[0].Version)
 }
 
+func TestUpgradeAppWithExistingAndDefaultAnswersForGitOpsInBatchMode(t *testing.T) {
+	testOptions := cmd_test_helpers.CreateAppTestOptions(true, t)
+	defer func() {
+		err := testOptions.Cleanup()
+		assert.NoError(t, err)
+	}()
+
+	// Needs console
+	console := tests.NewTerminal(t)
+	testOptions.CommonOptions.In = console.In
+	testOptions.CommonOptions.Out = console.Out
+	testOptions.CommonOptions.Err = console.Err
+
+	name, alias, version, err := testOptions.DirectlyAddAppToGitOps(map[string]interface{}{
+		"name": "testing",
+	})
+	assert.NoError(t, err)
+
+	envDir, err := testOptions.CommonOptions.EnvironmentsDir()
+	assert.NoError(t, err)
+	appDir := filepath.Join(envDir, testOptions.OrgName, testOptions.DevEnvRepoInfo.Name, name)
+
+	existingValues, err := ioutil.ReadFile(filepath.Join(appDir, helm.ValuesFileName))
+	assert.NoError(t, err)
+	assert.Equal(t, `name: testing
+`, string(existingValues))
+
+	// Now let's upgrade
+
+	newVersion, err := semver.Parse(version)
+	assert.NoError(t, err)
+	newVersion.Patch++
+	o := &cmd.UpgradeAppsOptions{
+		AddOptions: cmd.AddOptions{
+			CommonOptions: *testOptions.CommonOptions,
+		},
+		Version:              newVersion.String(),
+		Alias:                alias,
+		Repo:                 cmd_test_helpers.FakeChartmusuem,
+		GitOps:               true,
+		HelmUpdate:           true,
+		DevEnv:               testOptions.DevEnv,
+		ConfigureGitCallback: testOptions.ConfigureGitFn,
+	}
+	o.Args = []string{name}
+
+	helm_test.StubFetchChart(name, newVersion.String(),
+		cmd_test_helpers.FakeChartmusuem, &chart.Chart{
+			Metadata: &chart.Metadata{
+				Name:    name,
+				Version: newVersion.String(),
+			},
+			Files: []*google_protobuf.Any{
+				&google_protobuf.Any{
+					TypeUrl: "values.schema.json",
+					Value: []byte(`{
+  "$id": "https:/jenkins-x.io/tests/basicTypes.schema.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "description": "test values.yaml",
+  "type": "object",
+  "properties": {
+    "name": {
+      "type": "string"
+    },
+    "species": {
+      "type": "string",
+      "default": "human"
+    }
+  }
+}`),
+				},
+			},
+		}, testOptions.MockHelmer)
+
+	err = o.Run()
+	assert.NoError(t, err)
+	// Validate a PR was created
+	_, err = testOptions.FakeGitProvider.GetPullRequest(testOptions.OrgName, testOptions.DevEnvRepoInfo, 1)
+	assert.NoError(t, err)
+	// Validate the updated values.yaml
+	existingValues, err = ioutil.ReadFile(filepath.Join(appDir, helm.ValuesFileName))
+	assert.NoError(t, err)
+	assert.Equal(t, `name: testing
+species: human
+`, string(existingValues))
+}
+
+func TestUpgradeAppWithExistingAndDefaultAnswersForGitOps(t *testing.T) {
+	testOptions := cmd_test_helpers.CreateAppTestOptions(true, t)
+	defer func() {
+		err := testOptions.Cleanup()
```

Comment:
will that tear down any created directories, artefacts, etc, right?

Reply:
Seems quite a bit of work to set up a test. What's your experience with maintaining of these tests?

Reply:
> will that tear down any created directories, artefacts, etc, right?

Yes

Reply:
> Seems quite a bit of work to set up a test. What's your experience with maintaining of these tests?

The thing that is very hard to maintain is the expect based tests - they just randomly go wrong sometimes :-( The rest is basically ok - I've tried to put what I can in helper methods that generically does stuff - then the test can focus on logic - the one you comment on the console stuff should go into the helper really, but the rest is really actually test specific - adding an app to do an upgrade against, running the upgrade, mocking the chart repo (via the `StubFetchChart`), running the console, validating the PR was created correctly.